src/core/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,370 @@
+  );
+
+  useTheme();
+
+  /** Generates an memoized id for an option, given the index from original list */
+  const getOptionId = React.useCallback(
+    (index: number) =>
+      options[index].id ??
+      `${id}-option${options.findIndex(
+        ({ value }) => value === options[index].value,
+      )}`,
+    [options, id],
+  );
+
+  const memoizedItems = React.useMemo(
+    () =>
+      options.map(({ label, value, ...rest }, index) => (
+        <MenuItem
+          id={getOptionId(index)}
+          key={getOptionId(index)}
+          value={value}
+          role='option'
+          onClick={(value: T) => {
+            setSelectedValue(value);
+            setIsOpen(false);
+          }}
+          {...rest}
+        >
+          {label}
+        </MenuItem>
+      )),
+    [options, getOptionId],
+  );
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const menuRef = React.useRef<HTMLUListElement>(null);
+  const toggleButtonRef = React.useRef<HTMLSpanElement>(null);
+  const [isOpen, setIsOpen] = React.useState(false);
+
+  // Set min-width of menu to be same as input
+  const [minWidth, setMinWidth] = React.useState(0);
+  React.useEffect(() => {
+    if (inputRef.current) {
+      setMinWidth(inputRef.current.offsetWidth);
+    }
+  }, [isOpen]);
+
+  const [filteredOptions, setFilteredOptions] = React.useState(options);
+  const [focusedIndex, setFocusedIndex] = React.useState(() =>
+    options.findIndex((option) => value === option.value),
+  );
+
+  // Maintain internal selected value state synced with `value` prop
+  const [selectedValue, setSelectedValue] = React.useState<T | undefined>();
+  React.useEffect(() => {
+    setSelectedValue(value);
+  }, [value]);
+
+  // Controlled input value
+  const [inputValue, setInputValue] = React.useState<string>('');
+  const onInput = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setInputValue(event.target.value);
+      inputProps?.onChange?.(event);
+    },
+    [inputProps],
+  );
+
+  // Fire onChange callback and update inputValue every time selected value changes
+  React.useEffect(() => {
+    if (selectedValue != undefined) {
+      onChange?.(selectedValue);
+      setInputValue(
+        options.find(({ value }) => value === selectedValue)?.label ?? '',
+      );
+    }
+  }, [selectedValue, onChange, options]);
+
+  // Filter options and update focus when input value changes
+  React.useEffect(() => {
+    if (!isOpen) {
+      return;
+    }
+
+    // if input is empty or same as selected value, show the whole list
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    if (!inputValue || selectedOption?.label === inputValue) {
+      setFilteredOptions(options);
+      return;
+    }
+
+    const _filteredOptions =
+      filterFunction?.(options, inputValue) ??
+      options.filter((option) =>
+        option.label.toLowerCase().includes(inputValue?.trim().toLowerCase()),
```

Comment:
Not sure but can you make this also find the inputValue in option.value instead of just option.label? 

For example if I search for American Samoa by typing AS shouldn't it show as a result?

![image](https://user-images.githubusercontent.com/88331924/135676238-c54ef64c-fe0d-4ec2-bdbb-d5241cd40d1a.png)
![image](https://user-images.githubusercontent.com/88331924/135676248-2cb0b352-0dd8-4d6c-b46a-415120bd64f3.png)

Reply:
That would not work because `option.value` will not always be a string. But for that use case, I expose the `filterFunction` so it is still possible on the user's side.